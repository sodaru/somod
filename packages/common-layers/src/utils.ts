import { existsSync } from "fs";
import { mkdir, readdir, readFile, writeFile } from "fs/promises";
import { join } from "path";
import { chdir } from "process";
import { CommonLayers, layerLibraries } from "./index";
import rimraf from "rimraf";

const deleteDir = (dir: string): Promise<void> => {
  return new Promise((resolve, reject) => {
    rimraf(dir, err => {
      if (err) {
        reject(err);
      } else {
        resolve();
      }
    });
  });
};

/**
 * Generates layer structure and package.json(including dependencies) for all configured layers
 * layer structure will be like {base folder}/layers/{layer name}/nodejs/package.json
 */
export const generate = async () => {
  chdir(join(__dirname, ".."));
  await mkdir("./layers", { recursive: true });
  const rootPackageJsonStr = await readFile("./package.json", {
    encoding: "utf8"
  });

  const rootPackageJson = JSON.parse(rootPackageJsonStr);

  for (const layer in layerLibraries) {
    const dependencies = {};
    layerLibraries[layer]["libraries"].forEach((dependency: string) => {
      const dependencyVersion = rootPackageJson.peerDependencies[dependency];
      if (!dependencyVersion) {
        throw new Error(
          `Dependency ${dependency} in layer ${layer} is not found in root package.json's peerDependencies`
        );
      }
      dependencies[dependency] = dependencyVersion;
    });
    const packageJSON = {
      name: `slplayer-${layer.toLowerCase()}`,
      version: "1.0.0",
      description: "Function layer generated by @somod/common-layers",
      dependencies: dependencies
    };
    await mkdir(`./layers/${layer}/nodejs`, { recursive: true });
    await writeFile(
      `./layers/${layer}/nodejs/package.json`,
      JSON.stringify(packageJSON, null, 2)
    );
  }
};

/**
 * compares and removes duplicate libraries from passed layer w.r.t base layer
 * @param layerName layer name for comparing duplicate libraries
 */
export const removeBaseLayerLibraries = async (
  layerName: string
): Promise<void> => {
  chdir(join(__dirname, ".."));
  const baseLayerNodeModulesPath = `./layers/${CommonLayers.baseLayer}/nodejs/node_modules`;
  const layerNodeModulesPath = `./layers/${layerName}/nodejs/node_modules`;
  const dirs = await readdir(layerNodeModulesPath);
  const scopes: string[] = [];
  const libraries = dirs.filter(dir => {
    if (dir.startsWith("@")) {
      scopes.push(dir);
      return false;
    }
    return true;
  });

  await Promise.all(
    scopes.map(async scope => {
      const scopedLibraries = await readdir(`${layerNodeModulesPath}/${scope}`);
      libraries.push(...scopedLibraries.map(l => `${scope}/${l}`));
    })
  );

  await Promise.all(
    libraries.map(async lib => {
      if ([".bin", ".package-lock.json"].includes(lib)) {
        return;
      }
      if (existsSync(`${baseLayerNodeModulesPath}/${lib}`)) {
        const packageJsonStrList = await Promise.all([
          readFile(`${baseLayerNodeModulesPath}/${lib}/package.json`, {
            encoding: "utf8"
          }),
          readFile(`${layerNodeModulesPath}/${lib}/package.json`, {
            encoding: "utf8"
          })
        ]);
        const [baseLayerLibVersion, layerLibVersion] = packageJsonStrList.map(
          p => JSON.parse(p).version
        );
        if (baseLayerLibVersion == layerLibVersion) {
          await deleteDir(`${layerNodeModulesPath}/${lib}`);
        }
      }
    })
  );
};
