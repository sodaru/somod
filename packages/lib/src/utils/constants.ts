export const file_packageJson = "package.json";
export const file_packageLockJson = "package-lock.json";

export const file_dotenv = ".env";

export const file_tsConfigSomodJson = "tsconfig.somod.json";

export const file_nextConfigJs = "next.config.js";

export const key_somod = "somod";

export const key_type = "type";

export const key_module = "module";
export const key_typings = "typings";

export const key_sideEffects = "sideEffects";

export const key_main = "main";
export const key_jsnextMain = "jsnext:main";
export const key_files = "files";

export const key_serverlessSchema = "serverlessSchema";

export const key_somodPlugins = "somodPlugins";

export const path_nodeModules = "node_modules";

export const path_somodWorkingDir = ".somod";

export const path_build = "build";

export const path_ui = "ui";
export const path_lib = "lib";

export const path_pages = "pages";
export const path_pagesData = "pages-data";
export const path_public = "public";

export const file_index_js = "index.js";
export const file_index_dts = "index.d.ts";

export const path_git = ".git";

export const file_configYaml = "config.yaml";
export const file_configJson = "config.json";

export const file_extensionTs = "extension.ts";
export const file_extensionJs = "extension.js";

export const path_serverless = "serverless";
export const path_functions = "functions";
export const path_functionLayers = "functionLayers";
export const path_middlewares = "middlewares";
export const file_templateYaml = "template.yaml";
export const file_templateJson = "template.json";

export const file_samConfig = "samconfig.toml";

export const path_vercel = ".vercel";

export const cli_version_regex =
  /^([0-9]+)\.([0-9]+)\.([0-9]+)(?:-([0-9A-Za-z-]+(?:\.[0-9A-Za-z-]+)*))?(?:\+[0-9A-Za-z-]+)?$/;

export const namespace_page = "UI Page";

export const namespace_public = "UI Public Asset";

export const namespace_env_config = "UI Env Config";

export const namespace_public_runtime_config = "UI Public Runtime Config";

export const namespace_server_runtime_config = "UI Server Runtime Config";

export const namespace_api_gateway = "Serverless Api Gateway";

export const namespace_parameter = "Parameter";

export const namespace_output = "Serverless Output";

export const resourceType_Function = "AWS::Serverless::Function";
export const resourceType_FunctionLayer = "AWS::Serverless::LayerVersion";
export const resourceType_FunctionMiddleware =
  "SOMOD::Serverless::FunctionMiddleware";

export const custom_resource_prefix = "Custom::";

export const file_parametersYaml = "parameters.yaml";

export const file_parametersJson = "parameters.json";

export const key_parameter_overrides = "parameter_overrides";
