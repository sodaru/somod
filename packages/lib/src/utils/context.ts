import { createHash } from "crypto";
import { normalize } from "path";
import {
  IContext,
  IExtensionHandler,
  IModuleHandler,
  INamespaceHandler,
  IServerlessTemplateHandler
} from "somod-types";
import { ExtensionHandler } from "./extension/handler";
import { ModuleHandler } from "./module";
import { NamespaceHandler } from "./namespace";
import { ServerlessTemplateHandler } from "./serverless/serverlessTemplate/serverlessTemplate";

export class Context implements IContext {
  private static instance: IContext;

  private _dir: string;
  private _isUI: boolean;
  private _isServerless: boolean;
  private _isDebugMode: boolean;
  private _moduleHandler: IModuleHandler;
  private _extensionHandler: IExtensionHandler;
  private _namespaceHandler: INamespaceHandler;
  private _serverlessTemplateHandler: IServerlessTemplateHandler;

  private constructor() {
    // do nothing
  }

  static async getInstance(
    dir: string,
    isUI: boolean,
    isServerless: boolean,
    isDebugMode: boolean
  ) {
    if (this.instance === undefined) {
      const context = new Context();
      context._dir = normalize(dir);
      context._isUI = isUI;
      context._isServerless = isServerless;
      context._isDebugMode = isDebugMode;

      context._moduleHandler = await ModuleHandler.getInstance(context._dir);

      context._serverlessTemplateHandler =
        await ServerlessTemplateHandler.getInstance(context);

      context._extensionHandler = await ExtensionHandler.getInstance(
        context._moduleHandler
      );
      context._namespaceHandler = await NamespaceHandler.getInstance(context);

      this.instance = context;
    }
    return this.instance;
  }

  get dir() {
    return this._dir;
  }
  get moduleHandler() {
    return this._moduleHandler;
  }
  get extensionHandler(): IExtensionHandler {
    return this._extensionHandler;
  }
  get namespaceHandler() {
    return this._namespaceHandler;
  }
  get serverlessTemplateHandler() {
    return this._serverlessTemplateHandler;
  }
  get isUI() {
    return this._isUI;
  }
  get isServerless() {
    return this._isServerless;
  }
  get isDebugMode() {
    return this._isDebugMode;
  }
  getModuleHash(moduleName: string) {
    return createHash("sha256")
      .update(moduleName)
      .digest("hex")
      .substring(0, 8);
  }
}
