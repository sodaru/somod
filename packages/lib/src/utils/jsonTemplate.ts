import {
  intersection,
  isArray,
  isBoolean,
  isNull,
  isNumber,
  isPlainObject,
  isString
} from "lodash";
import {
  JSONType,
  JSONNode,
  JSONObjectType,
  KeywordValidator,
  KeywordProcessor,
  KeywordReplacement,
  KeywordObjectReplacement,
  KeywordKeywordReplacement
} from "somod-types";

export const parseJson = (json: JSONType): JSONNode => {
  const navigator = (json: JSONType, parent?: JSONNode["parent"]): JSONNode => {
    if (isArray(json)) {
      const itemNodes: JSONNode[] = [];
      const thisNode: JSONNode = {
        type: "array",
        items: itemNodes,
        parent: parent
      };
      Object.freeze(thisNode);
      json.forEach((item, i) => {
        const itemNode = navigator(item, {
          key: i,
          node: thisNode
        });
        itemNodes.push(itemNode);
      });
      Object.freeze(itemNodes);
      return thisNode;
    } else if (isPlainObject(json)) {
      const propertyNodes: Record<string, JSONNode> = {};
      const thisNode: JSONNode = {
        type: "object",
        properties: propertyNodes,
        parent: parent
      };
      Object.freeze(thisNode);
      Object.keys(json).forEach(property => {
        const propertyNode = navigator(json[property], {
          key: property,
          node: thisNode
        });
        propertyNodes[property] = propertyNode;
      });
      Object.freeze(propertyNodes);
      return thisNode;
    } else if (
      isString(json) ||
      isBoolean(json) ||
      isNumber(json) ||
      isNull(json)
    ) {
      const thisNode: JSONNode = {
        type: "primitive",
        value: json,
        parent: parent
      };
      Object.freeze(thisNode);
      return thisNode;
    } else {
      throw new Error(`Unknown type ${json} ${parent}`);
    }
  };
  return navigator(json);
};

export const getPath = (jsonNode: JSONNode): (string | number)[] => {
  const path: (string | number)[] = [];
  let currentNode = jsonNode;
  while (currentNode) {
    if (currentNode.parent) {
      const key = currentNode.parent.key;
      path.unshift(key);
      currentNode = currentNode.parent.node;
    } else {
      currentNode = null;
    }
  }
  return path;
};

export const constructJson = (jsonNode: JSONNode): JSONType => {
  if (jsonNode.type == "primitive") {
    return jsonNode.value;
  } else if (jsonNode.type == "array") {
    return jsonNode.items.map(item => constructJson(item));
  } else if (jsonNode.type == "object") {
    return Object.fromEntries(
      Object.keys(jsonNode.properties).map(property => [
        property,
        constructJson(jsonNode.properties[property])
      ])
    );
  }
};

export class JSONTemplateError extends Error {
  private _node: JSONNode;
  private _error: Error;

  constructor(node: JSONNode, error: Error) {
    const message = `Error at ${getPath(node).join(".")} : ${error.message}`;
    super(message);
    this._node = node;
    this._error = error;

    // Set the prototype explicitly.
    Object.setPrototypeOf(this, new.target.prototype);
  }

  get path() {
    return getPath(this._node);
  }

  get node() {
    return this._node;
  }

  get error() {
    return this._error;
  }
}

export const validateKeywords = async (
  jsonNode: JSONNode,
  keywordValidators: Record<string, KeywordValidator>
): Promise<JSONTemplateError[]> => {
  const errors: JSONTemplateError[] = [];
  const keywords = Object.keys(keywordValidators);
  const navigator = async (jsonNode: JSONNode) => {
    if (jsonNode.type == "array") {
      await Promise.all(
        jsonNode.items.map(async item => {
          await navigator(item);
        })
      );
    } else if (jsonNode.type == "object") {
      const properties = Object.keys(jsonNode.properties);
      await Promise.all(
        properties.map(async property => {
          await navigator(jsonNode.properties[property]);
        })
      );

      // validation
      const keywordsInThisObject = intersection(properties, keywords);

      await Promise.all(
        keywordsInThisObject.map(async keyword => {
          const keywordErrorsPromise = keywordValidators[keyword](
            keyword,
            jsonNode,
            constructJson(jsonNode.properties[keyword])
          );
          let keywordErrors = keywordErrorsPromise as Error[];
          if (
            typeof (keywordErrorsPromise as Promise<Error[]>)?.then ==
            "function"
          ) {
            keywordErrors = await keywordErrorsPromise;
          }

          errors.push(
            ...keywordErrors.map(e => new JSONTemplateError(jsonNode, e))
          );
        })
      );
    }
  };
  await navigator(jsonNode);
  return errors;
};

class ReplaceObjectAtLevelError extends Error {
  private _node: JSONNode;
  private _level: number;
  private _replacement: JSONType;

  constructor(node: JSONNode, level: number, replacement: JSONType) {
    super(`Object replacement at level ${level} was not possible`);
    this._node = node;
    this._level = level;
    this._replacement = replacement;

    // Set the prototype explicitly.
    Object.setPrototypeOf(this, new.target.prototype);
  }

  next() {
    this._level--;
    return this;
  }

  get node() {
    return this._node;
  }

  get level() {
    return this._level;
  }

  get replacement() {
    return this._replacement;
  }
}

export const processKeywords = async (
  jsonNode: JSONNode,
  keywordProcessors: Record<string, KeywordProcessor>
): Promise<JSONType> => {
  const keywords = Object.keys(keywordProcessors);

  const navigator = async (jsonNode: JSONNode): Promise<JSONType> => {
    if (jsonNode.type == "primitive") {
      return jsonNode.value;
    } else if (jsonNode.type == "array") {
      const jsonArray = await Promise.all(
        jsonNode.items.map(item => navigatorWithReplacementHandler(item))
      );
      return jsonArray;
    } else if (jsonNode.type == "object") {
      const properties = Object.keys(jsonNode.properties);
      const jsonObject = Object.fromEntries(
        await Promise.all(
          properties.map(async property => [
            property,
            await navigatorWithReplacementHandler(jsonNode.properties[property])
          ])
        )
      ) as JSONObjectType;

      // replacement
      const keywordsInThisObject = intersection(properties, keywords);
      const objectReplacements: Record<string, KeywordObjectReplacement> = {};
      const keywordReplacements: Record<string, KeywordKeywordReplacement> = {};
      await Promise.all(
        keywordsInThisObject.map(async keyword => {
          try {
            const replacerPromise = keywordProcessors[keyword](
              keyword,
              jsonNode,
              jsonObject[keyword]
            );
            let replacer = replacerPromise as KeywordReplacement;
            if (
              typeof (replacerPromise as Promise<KeywordReplacement>)?.then ==
              "function"
            ) {
              replacer = await replacerPromise;
            }
            if (replacer.type == "object") {
              objectReplacements[keyword] = replacer;
            } else {
              keywordReplacements[keyword] = replacer;
            }
          } catch (e) {
            throw new JSONTemplateError(jsonNode, e);
          }
        })
      );

      const numberOfObjectReplacements = Object.keys(objectReplacements).length;
      const numberOfKeywordReplacements =
        Object.keys(keywordReplacements).length;

      if (
        numberOfObjectReplacements > 1 ||
        (numberOfObjectReplacements == 1 && numberOfKeywordReplacements > 0)
      ) {
        throw new JSONTemplateError(
          jsonNode,
          new Error(
            `Object replacement can not be combined with other object/keyword replacements. The keywords are ${keywordsInThisObject.join(
              ", "
            )}`
          )
        );
      }

      if (numberOfObjectReplacements == 1) {
        const replacement = Object.values(objectReplacements)[0];
        if (replacement.level > 0) {
          throw new ReplaceObjectAtLevelError(
            jsonNode,
            replacement.level,
            replacement.value
          );
        }
        return Object.values(objectReplacements)[0].value;
      } else {
        Object.keys(keywordReplacements).forEach(k => {
          delete jsonObject[k];
          const replacement = keywordReplacements[k].value;
          Object.keys(replacement).forEach(newKey => {
            jsonObject[newKey] = replacement[newKey];
          });
        });
        return jsonObject;
      }
    }
  };

  const navigatorWithReplacementHandler = async (
    jsonNode: JSONNode
  ): Promise<JSONType> => {
    try {
      return await navigator(jsonNode);
    } catch (e) {
      if (e instanceof ReplaceObjectAtLevelError) {
        if (e.level > 0) {
          throw e.next();
        } else {
          return e.replacement;
        }
      } else {
        throw e;
      }
    }
  };

  try {
    return await navigatorWithReplacementHandler(jsonNode);
  } catch (e) {
    if (e instanceof ReplaceObjectAtLevelError) {
      throw new JSONTemplateError(e.node, e);
    } else {
      throw e;
    }
  }
};
