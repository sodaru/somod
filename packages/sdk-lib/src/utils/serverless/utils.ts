import { createHash } from "crypto";
import { cloneDeep, isArray, isPlainObject, isUndefined } from "lodash";
import { KeywordAll, SLPKeyword, SLPTemplate } from "./types";

const getSLPKeywordPaths = (chunk: unknown): SLPTemplate["keywordPaths"] => {
  const keywordPaths = Object.fromEntries(
    KeywordAll.map(keyword => [keyword, []])
  ) as SLPTemplate["keywordPaths"];

  if (isPlainObject(chunk)) {
    KeywordAll.forEach(keyword => {
      if (!isUndefined(chunk[keyword])) {
        keywordPaths[keyword].push([]);
      }
    });

    Object.keys(chunk).forEach(key => {
      const childKeywordPaths = getSLPKeywordPaths(chunk[key]);
      KeywordAll.forEach(keyword => {
        childKeywordPaths[keyword].forEach((_keywordPaths: string[]) => {
          _keywordPaths.unshift(key);
          keywordPaths[keyword].push(_keywordPaths);
        });
      });
    });
  } else if (isArray(chunk)) {
    chunk.forEach((arrayItem, index) => {
      const childKeywordPaths = getSLPKeywordPaths(arrayItem);
      KeywordAll.forEach(keyword => {
        childKeywordPaths[keyword].forEach((_keywordPaths: string[]) => {
          _keywordPaths.unshift(index + "");
          keywordPaths[keyword].push(_keywordPaths);
        });
      });
    });
  }
  return keywordPaths;
};

export const updateKeywordPathsInSLPTemplate = (
  slpTemplate: SLPTemplate
): void => {
  const slpKeywords = getSLPKeywordPaths(slpTemplate.original.Resources);
  slpTemplate.keywordPaths = slpKeywords;
};

/**
 * returns the value present in the path (usually Keywords path), and it always looks in original
 * @param slpTemplate : single slp tamplate
 * @param path : json path in array format (usually Keywords path)
 * @returns value found inside the json path
 *          for example : { "SLP::Ref": { module:"@sodaru/slp", "resource": "BaseRestApi" } }
 */
export const getSLPKeyword = <T extends SLPKeyword>(
  slpTemplate: SLPTemplate,
  path: string[]
): T => {
  // always return the keyword value from original
  let keyword: unknown = slpTemplate.original.Resources;

  path.forEach(pathSegment => {
    keyword = keyword[pathSegment];
  });
  return cloneDeep(keyword) as T;
};

export const replaceSLPKeyword = (
  slpTemplate: SLPTemplate,
  path: string[],
  newValue: unknown
): void => {
  let keyword: unknown = slpTemplate.Resources;
  const _path = [...path];
  const lastPathSegment = _path.pop();
  _path.forEach(pathSegment => {
    keyword = keyword[pathSegment];
  });
  keyword[lastPathSegment] = newValue;
};

const hashModuleName = (str: string): string => {
  return createHash("sha256").update(str).digest("hex").substring(0, 8);
};

export const getSAMParameterName = (
  moduleName: string,
  slpParameterName: string
): string => {
  return "p" + hashModuleName(moduleName) + slpParameterName;
};

export const getSAMResourceLogicalId = (
  moduleName: string,
  slpResourceId: string
): string => {
  return "r" + hashModuleName(moduleName) + slpResourceId;
};

export const getSAMResourceName = (
  moduleName: string,
  slpResourceName: string
): unknown => {
  return {
    "Fn::Sub": [
      "slp${stackId}${moduleHash}${slpResourceName}",
      {
        stackId: {
          "Fn::Select": [2, { "Fn::Split": ["/", { Ref: "AWS::StackId" }] }]
        },
        moduleHash: hashModuleName(moduleName),
        slpResourceName
      }
    ]
  };
};
