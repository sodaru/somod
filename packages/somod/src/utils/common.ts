import { Command } from "nodejs-cli-runner";

export type SOMODCommandTypeOptions = {
  ui: boolean;
  serverless: boolean;
};

export type DebugModeOptions = {
  debug: boolean;
};

export const addSOMODCommandTypeOptions = (command: Command) => {
  command.option("--ui", "only ui");
  command.option("--serverless", "only serverless");
};

export const addDebugOptions = (command: Command) => {
  command.option("-d, --debug", "Enable Debug mode");
};

export const getSOMODCommandTypeOptions = (
  options: SOMODCommandTypeOptions
): SOMODCommandTypeOptions => {
  const ui = !options.serverless || options.ui;
  const serverless = !options.ui || options.serverless;
  return { ui, serverless };
};
