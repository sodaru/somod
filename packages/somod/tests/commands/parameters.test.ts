import parametersCommand from "../../src/commands/parameters";
import { IContext } from "somod-types";
import * as lib from "somod-lib";
import { mockedFunction } from "../utils";
import { addCommonOptions } from "nodejs-cli-runner";

jest.mock("somod-lib", () => {
  const original = jest.requireActual("somod-lib");
  const mocked = {};
  for (const key in original) {
    mocked[key] =
      typeof original[key] === "function" ? jest.fn() : original[key];
  }
  return {
    __esModule: true,
    ...mocked
  };
});

describe("Test command parameters", () => {
  let output: unknown[][] = [];

  // eslint-disable-next-line no-console
  const originalConsoleLog = console.log;

  const functionsToBeMocked = [
    "initializeContext",
    "updateParametersFromSAM",
    "validateParameterValues"
  ];

  const mock = (input: Record<string, unknown>) => {
    for (const key of functionsToBeMocked) {
      // eslint-disable-next-line import/namespace
      mockedFunction(lib[key]).mockImplementation(async (...args) => {
        output.push([key, ...args]);
        return input[key];
      });
    }
    mockedFunction(lib.findRootDir).mockImplementation((...args) => {
      output.push(["findRootDir", ...args]);
      return "/root/dir";
    });
  };

  beforeAll(() => {
    addCommonOptions(parametersCommand);
  });

  beforeEach(() => {
    parametersCommand.commands[0].setOptionValue("verbose", undefined);
    parametersCommand.commands[0].setOptionValue("debug", undefined);
    parametersCommand.commands[0].setOptionValue("stackName", undefined);
    parametersCommand.commands[1].setOptionValue("verbose", undefined);
    parametersCommand.commands[1].setOptionValue("debug", undefined);

    // eslint-disable-next-line no-console
    console.log = originalConsoleLog;
    for (const key of functionsToBeMocked) {
      // eslint-disable-next-line import/namespace
      mockedFunction(lib[key]).mockReset();
    }
    output = [];

    // eslint-disable-next-line no-console
    console.log = jest.fn();
  });

  describe("update subcommand", () => {
    test("without args", async () => {
      mock({
        initializeContext: {
          dir: "/root/dir"
        } as IContext
      });
      await parametersCommand.parseAsync(["update"], {
        from: "user"
      });
      expect(output).toMatchSnapshot();
    });

    test("with verbose", async () => {
      mock({
        initializeContext: {
          dir: "/root/dir"
        } as IContext
      });
      await parametersCommand.parseAsync(["update", "-v"], {
        from: "user"
      });
      expect(output).toMatchSnapshot();
    });

    test("with debug mode", async () => {
      mock({
        initializeContext: {
          dir: "/root/dir"
        } as IContext
      });
      await parametersCommand.parseAsync(["update", "-d"], {
        from: "user"
      });
      expect(output).toMatchSnapshot();
    });

    test("with stackname", async () => {
      mock({
        initializeContext: {
          dir: "/root/dir"
        } as IContext
      });
      await parametersCommand.parseAsync(
        ["update", "--stack-name", "my-stack"],
        {
          from: "user"
        }
      );
      expect(output).toMatchSnapshot();
    });

    test("with all options", async () => {
      mock({
        initializeContext: {
          dir: "/root/dir"
        } as IContext
      });
      await parametersCommand.parseAsync(
        ["update", "-v", "-d", "--stack-name", "my-stack"],
        {
          from: "user"
        }
      );
      expect(output).toMatchSnapshot();
    });
  });

  describe("validate subcommand", () => {
    test("without args", async () => {
      mock({
        initializeContext: {
          dir: "/root/dir"
        } as IContext
      });
      await parametersCommand.parseAsync(["validate"], {
        from: "user"
      });
      expect(output).toMatchSnapshot();
    });

    test("with verbose", async () => {
      mock({
        initializeContext: {
          dir: "/root/dir"
        } as IContext
      });
      await parametersCommand.parseAsync(["validate", "-v"], {
        from: "user"
      });
      expect(output).toMatchSnapshot();
    });

    test("with debug mode", async () => {
      mock({
        initializeContext: {
          dir: "/root/dir"
        } as IContext
      });
      await parametersCommand.parseAsync(["validate", "-d"], {
        from: "user"
      });
      expect(output).toMatchSnapshot();
    });

    test("with all options", async () => {
      mock({
        initializeContext: {
          dir: "/root/dir"
        } as IContext
      });
      await parametersCommand.parseAsync(["validate", "-v", "-d"], {
        from: "user"
      });
      expect(output).toMatchSnapshot();
    });
  });
});
