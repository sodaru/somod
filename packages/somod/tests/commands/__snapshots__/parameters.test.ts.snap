// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test command parameters update subcommand with all options 1`] = `
Array [
  Array [
    "findRootDir",
  ],
  Array [
    "initializeContext",
    "/root/dir",
    false,
    true,
    true,
  ],
  Array [
    "updateParametersFromSAM",
    Object {
      "dir": "/root/dir",
    },
    "my-stack",
  ],
]
`;

exports[`Test command parameters update subcommand with debug mode 1`] = `
Array [
  Array [
    "findRootDir",
  ],
  Array [
    "initializeContext",
    "/root/dir",
    false,
    true,
    true,
  ],
  Array [
    "updateParametersFromSAM",
    Object {
      "dir": "/root/dir",
    },
    undefined,
  ],
]
`;

exports[`Test command parameters update subcommand with stackname 1`] = `
Array [
  Array [
    "findRootDir",
  ],
  Array [
    "initializeContext",
    "/root/dir",
    false,
    true,
    undefined,
  ],
  Array [
    "updateParametersFromSAM",
    Object {
      "dir": "/root/dir",
    },
    "my-stack",
  ],
]
`;

exports[`Test command parameters update subcommand with verbose 1`] = `
Array [
  Array [
    "findRootDir",
  ],
  Array [
    "initializeContext",
    "/root/dir",
    false,
    true,
    undefined,
  ],
  Array [
    "updateParametersFromSAM",
    Object {
      "dir": "/root/dir",
    },
    undefined,
  ],
]
`;

exports[`Test command parameters update subcommand without args 1`] = `
Array [
  Array [
    "findRootDir",
  ],
  Array [
    "initializeContext",
    "/root/dir",
    false,
    true,
    undefined,
  ],
  Array [
    "updateParametersFromSAM",
    Object {
      "dir": "/root/dir",
    },
    undefined,
  ],
]
`;

exports[`Test command parameters validate subcommand with all options 1`] = `
Array [
  Array [
    "findRootDir",
  ],
  Array [
    "initializeContext",
    "/root/dir",
    false,
    false,
    true,
  ],
  Array [
    "validateParameterValues",
    Object {
      "dir": "/root/dir",
    },
  ],
]
`;

exports[`Test command parameters validate subcommand with debug mode 1`] = `
Array [
  Array [
    "findRootDir",
  ],
  Array [
    "initializeContext",
    "/root/dir",
    false,
    false,
    true,
  ],
  Array [
    "validateParameterValues",
    Object {
      "dir": "/root/dir",
    },
  ],
]
`;

exports[`Test command parameters validate subcommand with verbose 1`] = `
Array [
  Array [
    "findRootDir",
  ],
  Array [
    "initializeContext",
    "/root/dir",
    false,
    false,
    undefined,
  ],
  Array [
    "validateParameterValues",
    Object {
      "dir": "/root/dir",
    },
  ],
]
`;

exports[`Test command parameters validate subcommand without args 1`] = `
Array [
  Array [
    "findRootDir",
  ],
  Array [
    "initializeContext",
    "/root/dir",
    false,
    false,
    undefined,
  ],
  Array [
    "validateParameterValues",
    Object {
      "dir": "/root/dir",
    },
  ],
]
`;
